### DISCUSSION TITLE
Proposal: Implementing Delete confirmation using Livewire Inbuilt Confirm Directive

### DISCUSSION THREAD

**Discussion by hemant-kr-meena** [Power-Components/livewire-powergrid#1585] [Ideas]
## Proposal:
Introduce macros named `confirm` and `confirmPrompt` in the Button class. These macros will simplify adding confirmation and prompt dialogs, ensuring consistency and preventing accidental deletions or crucial actions.

### Simple Confirm Popup
```php
Button::macro('confirm', function (string $message) {
    $this->dynamicProperties['confirm'] = [
        "component" => "button",
        "attribute" => "wire:confirm",
        "value"     => $message,
    ];

    return $this;
});
```
<img width="361" alt="confirm" src="https://github.com/Power-Components/livewire-powergrid/assets/38046347/065fe796-9a41-4634-97e1-0fa6ec8501b7">

### Confirm Prompt Popup with input

```php
Button::macro('confirmPrompt', function (string $message,string $confirm_value) {
    $this->dynamicProperties['confirmPrompt'] = [
        "component" => "button",
        "attribute" => "wire:confirm.prompt",
        "value"     => "$message |$confirm_value",
    ];

    return $this;
});
```
<img width="412" alt="confrim-prompt" src="https://github.com/Power-Components/livewire-powergrid/assets/38046347/c30e33cf-d650-407e-a583-e969a9cb2f9e">

**Comment by SuGDevelopers** [comment]
any implementation or how I could have this change in my project?

**Reply by hemant-kr-meena** [reply]
Yes in my recent project, I implement this logic to confirm before taking action.
As we know we use macros for additional functionality to add id and trigger dispatch events. In the same way, we can add wire:confirm attribute to buttons.
[Read more Livewire wire:confirm](https://livewire.laravel.com/docs/wire-confirm)


In my case, I add some new macros in AppServiceProvider for buttons and call then like other methods.


**Approach 1**

```php
Button::macro('confirm', function (string $message) {
    $this->dynamicProperties['confirm'] = [
        "component" => "button",
        "attribute" => "wire:confirm",
        "value"     => $message,
    ];

    return $this;
});
````

**Approach 2**

```php
Button::macro('confirmPrompt', function (string $message,string $confirm_value) {
    $this->dynamicProperties['confirmPrompt'] = [
        "component" => "button",
        "attribute" => "wire:confirm.prompt",
        "value"     => "$message |$confirm_value",
    ];

    return $this;
});

```

 
If You agree with suggestions, I will create a pull request with these functions. We can merge both functions and achieve the same result using a single macro. But by defining both methods we help users choose which option they want.

Tell me what you think.


<img width="837" alt="Screenshot (115)" src="https://github.com/Power-Components/livewire-powergrid/assets/38046347/443938e2-0194-45c2-8298-547bfeed7f81">


The results are the same as above images.

**Reply by hemant-kr-meena** [reply]
User can call them like



```php
Button::add('edit')->id()->confirm("Are you sure you want to delete ?")
```

**Reply by SuGDevelopers** [reply]
I have set it as a trait to be used in all my tables... thanks for the contribution... it works super well

**Reply by hemant-kr-meena** [reply]
I'm glad that it's helpful for you.

**Comment by hemant-kr-meena** [comment]
feature added in #1599
